### PR TITLE
extend serialization for tensor metadata

### PR DESCRIPTION
**Pull request by fakeYan**
Fixes #ISSUE_NUMBER
Add the serialization logic of backend metadata to the serialization of tensor, which is implemented through custom registration functions.

In #97429 , the structure backendMeta is provided in TensorImpl, and we think that this part of information may also need to be serialized for custom.

cc @mruberry @mikaylagawarecki

### CODE REVIEW

**Comment by fakeYan** [comment]
Hi @ezyang,

I really feel your hesitation and the risks you think exist in #98733，I try to use a new idea to break the current blocking.

As you can see, I changed the torch_check for the metadata dictionary to torch_warning. Doing so can greatly improve scalability.

The specific usage is as follows：

 1. Firstly, we use monkey patch to override the `get_tensor_metadata(tensor)` and `set_tensor_metadata(tensor, metadata)`, In this way, this part of the serialization logic can be completely implemented by the customer. Obviously, when the user uses the extension package, it can be saved and loaded correctly.
 2. When a .pt file has been stored and cross-platform, deserialization can only rely on the torch package. At this time, it is required that the torch package cannot directly report an error and hang up when it encounters an unknown key-value pair in the metadata dict. It only needs to parse the key-value pairs it knows.

Do you think such a change is acceptable? I think the risk of such a change is much less.

**Comment by fakeYan** [comment]
But I still think the monkey patch on the extension plug-in side is not the best solution, maybe we can not close pr #98733 directly and wait for other reviewer to review, I'm actively looking for help in the slack community.

**Comment by ezyang** [comment]
I'm not sure why you need this change then. If you are monkeypatching, can't you change the monkeypatch to remove these keys from the dict before passing them down below?

**Comment by fakeYan** [comment]
Hi @ezyang，

Very embarrassing, I don't know what wrong operation I did, closed pr #98733 and can't reopen it，can only create new pr

So I put the changed code here.

**Comment by fakeYan** [comment]
> I think this patch only allows a concrete backend metadata implementation to contain boolean fields, which would not be sufficient in our case. 

may be we can make metadata param from `unordered_map<std::string, bool>` to `unordered_map<std::string, int>`，I think such extended expression capacity must be sufficient

> Secondly at minimum the concrete backend type should be dependent on the device. I mean that getter/setted registry could be a mapping from device type to a pair of functions. Instead of hard-coding a check for `PrivateUse1` device it would be more generic to check if the registry contains backend metadata serializer/deserializer for a specific device type. For instance we are using the `HPU` key.

I think this part of the extension is very easy to solve, of course, we can not write a fixed backend, we can judge by parsing, or we can add conditions.

File: torch/csrc/jit/serialization/pickler.cpp [hunk 1]
```diff
@@ -801,4 +801,87 @@ bool checkHasValidSetGetState(const std::shared_ptr<c10::ClassType>& cls) {
...
 }
 
+// Declare BackendMeta serialization and deserialization function pointer types.
+using BackendMetaPtr =
+    void (*)(const at::Tensor&, std::unordered_map<std::string, int>&);
+
+// map to save function pointer for BackendMeta serialization.
+// key is the DeviceType, value is std::pair obj. 
+// value.first represent get function and value.seconde represent set function
+static std::unordered_map<int, std::pair<void*, void*>> serialization_map;
```
Comment (fakeYan) [inline torch/csrc/jit/serialization/pickler.cpp:811]: >OK, having understood the approach, I am more positive about this. Please stop using a string hash map with get/set fields though.

This variable `serialization_map` map the backend and the registered two serialization functions

File: torch/csrc/jit/serialization/pickler.cpp [hunk 2]
```diff
@@ -801,4 +801,87 @@ bool checkHasValidSetGetState(const std::shared_ptr<c10::ClassType>& cls) {
   return true;
 }
 
+// Declare BackendMeta serialization and deserialization function pointer types.
+using BackendMetaPtr =
```
Comment (fakeYan) [inline torch/csrc/jit/serialization/pickler.cpp:805]: `BackendMetaPtr ` is the function pionter form，both for serialization and deserialization

File: torch/csrc/jit/serialization/pickler.cpp [hunk 3]
```diff
@@ -801,4 +801,87 @@ bool checkHasValidSetGetState(const std::shared_ptr<c10::ClassType>& cls) {
...
+  if (t.is_conj()) {
+    metadata["conj"] = 1;
+  }
+  if (t.is_neg()) {
+    metadata["neg"] = 1;
+  }
+  // Only add BackendMetaData for custom backend if the function pointer is
+  // registered.
+  int device_type = static_cast<int>(t.device().type());
+  if (serialization_map.find(device_type) != serialization_map.end()) {
```
Comment (fakeYan) [inline torch/csrc/jit/serialization/pickler.cpp:844]: Here, the tensor device and the registered device are judged. When the two can be matched, this logic can be reached. 
Removed the logic of judging that device only is equal to PrivateUse1，so HPU can also use it.

**Comment by fakeYan** [comment]
Snipaste_2023-04-21_17-12-31 (2)

**Comment by fakeYan** [comment]
Hello @albanD and @ezyang, 

Please see if the modified serialization extension is ok in this time.

File: torch/csrc/jit/serialization/pickler.cpp [hunk 4]
```diff
@@ -801,4 +801,92 @@ bool checkHasValidSetGetState(const std::shared_ptr<c10::ClassType>& cls) {
...
 }
 
+// Declare BackendMeta serialization and deserialization function pointer types.
+using BackendMetaPtr =
+    void (*)(const at::Tensor&, std::unordered_map<std::string, bool>&);
+
+// map to save function pointer for BackendMeta serialization.
+// key is the DeviceType, value is std::pair obj.
+// value.first represent get function and value.seconde represent set function
+static std::unordered_map<int, std::pair<void*, void*>> serialization_map;
```
Comment (ezyang) [inline torch/csrc/jit/serialization/pickler.cpp:811]: DeviceType is an enum, so I think you can store it directly, without conversion to int. Also, you have the function pointer type on line 805, no need to cast the function pointer to void*.

File: torch/csrc/jit/serialization/pickler.cpp [hunk 5]
```diff
@@ -801,4 +801,92 @@ bool checkHasValidSetGetState(const std::shared_ptr<c10::ClassType>& cls) {
...
+
+// Register function pointer of Tensor BackendMetadata for serialization.
+void TensorBackendMetaRegistry(
+    c10::DeviceType t,
+    void* get_fptr,
+    void* set_fptr) {
```
Comment (ezyang) [inline torch/csrc/jit/serialization/pickler.cpp:817]: Then this can be typed too

**Comment by ezyang** [comment]
I think my general commentary is (1) make the code more type safe and (2) I don't want people to be able to register extensions to CPU device. So maybe whitelist a few allowed keys, or possible even only make this work for PrivateUse keys (don't even need an unordered map; an array would probably be fine.)

**Comment by fakeYan** [comment]
> I think my general commentary is (1) make the code more type safe and (2) I don't want people to be able to register extensions to CPU device. So maybe whitelist a few allowed keys, or possible even only make this work for PrivateUse keys (don't even need an unordered map; an array would probably be fine.)

I understand your thinking, I will modify accordingly.

**Review by ezyang — APPROVED** [review]

**Comment by fakeYan** [comment]
Hello @ezyang，

I found out that my changes were reverted，but I am not very clear about the reason "breaking internal builds" for the current fallback.

because my changes will affect some test cases or something, how should I continue to modify my pr and reopen

**Comment by atalman** [comment]
@fakeYan Hello here are some of the errors I see on our internal tests:

```
[2023-05-05T12:54:17.740-07:00] ld.lld: error: undefined symbol: torch::jit::setTensorMetadata(at::Tensor const&, c10::Dict<c10::IValue, c10::IValue>)
[2023-05-05T12:54:17.957-07:00] Action failed: fbsource//xplat/caffe2/fb/lite_predictor:lite_predictor_backend_demo (ovr_config//platform/linux:x86_64-fbcode-platform010-clang#ab2df4601d4b9022) (cxx_link_executable)
[CONTEXT] [2023-05-05T12:54:17.957-07:00] Remote command returned non-zero exit code 1
[CONTEXT] [2023-05-05T12:54:17.957-07:00] Remote action, reproduce with: `frecli cas download-action 63ddafe5be97afe8b233ba77010fc72ccf6457a5:94`
[2023-05-05T12:54:17.957-07:00] ld.lld: error: undefined symbol: torch::jit::setTensorMetadata(at::Tensor const&, c10::Dict<c10::IValue, c10::IValue>)
```

**Comment by fakeYan** [comment]
Hi @ezyang，

I found that the current ut is down. Because in #99882, a function is wrongly defined and called.

Snipaste_2023-05-11_20-23-01

I fix it in this pr

**Comment by fakeYan** [comment]
> @fakeYan Hello here are some of the errors I see on our internal tests:

> ```
> [2023-05-05T12:54:17.740-07:00] ld.lld: error: undefined symbol: torch::jit::setTensorMetadata(at::Tensor const&, c10::Dict<c10::IValue, c10::IValue>)
> [2023-05-05T12:54:17.957-07:00] Action failed: fbsource//xplat/caffe2/fb/lite_predictor:lite_predictor_backend_demo (ovr_config//platform/linux:x86_64-fbcode-platform010-clang#ab2df4601d4b9022) (cxx_link_executable)
> [CONTEXT] [2023-05-05T12:54:17.957-07:00] Remote command returned non-zero exit code 1
> [CONTEXT] [2023-05-05T12:54:17.957-07:00] Remote action, reproduce with: `frecli cas download-action 63ddafe5be97afe8b233ba77010fc72ccf6457a5:94`
> [2023-05-05T12:54:17.957-07:00] ld.lld: error: undefined symbol: torch::jit::setTensorMetadata(at::Tensor const&, c10::Dict<c10::IValue, c10::IValue>)
> ```

I interrupted the internal construction before, I think it is because the `inline `keyword was killed, I will add this part of the information now, and the construction should not be interrupted

**Comment by izaitsevfb** [comment]
@fakeYan, I'm sorry, but I had to revert your PR as it's breaking internal build at Meta with the following linker error:
```
stderr: ld.lld: error: undefined symbol: torch::jit::GetBackendMetaSerialization()
>>> referenced by pickler.h:368 (buck-out/../xplat/caffe2/torch/csrc/jit/serialization/pickler.h:368)
>>>               buck-out/gen/6cb65395/xplat/caffe2/torch_mobile_deserialize_commonAndroid#android-x86,compile-pic-unpickler.cpp.o5658cf0d/torch/csrc/jit/serialization/unpickler.cpp.o:(torch::jit::setTensorMetadata(at::Tensor const&, std::__ndk1::unordered_map<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char>>, bool, std::__ndk1::hash<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char>>>, std::__ndk1::equal_to<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char>>>, std::__ndk1::allocator<std::__ndk1::pair<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char>> const, bool>>>))
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
Unfortunately, I'm not sure how to repro this in OSS, but please let me know if I can be of any assistance.

**Comment by huydhn** [comment]
@fakeYan could you help rebase and resolve the conflict in the PR?  I can then import it as an internal Meta diff to see how it goes

**Comment by fakeYan** [comment]
> @fakeYan could you help rebase and resolve the conflict in the PR? I can then import it as an internal Meta diff to see how it goes

Of course, I will deal with the conflict as soon as possible

**Comment by huydhn** [comment]
It turns out that the test you need is in `ciflow/periodic`.  I have attached this label and BUCK jobs now fail accordingly https://hud.pytorch.org/pr/pytorch/pytorch/99808#13908933159

**Comment by fakeYan** [comment]
Hi @huydhn,

I found this problem with my code check (probably because I rebase the calculation record), now how do I pass this check, can you help me.
image

**Comment by heidongxianhua** [comment]
@huydhn   sorry to bother you, we have sloved the CI failure, could you please merge this pr again?

**Comment by huydhn** [comment]
From the internal signals, I think there is an discrepancy build issue where `-Werror,-Wglobal-constructors` is set:

```
torch/csrc/jit/serialization/pickler.h:305:44: error: declaration requires a global destructor [-Werror,-Wglobal-constructors]
[CONTEXT] static std::unordered_set<c10::DeviceType> DeviceTypeAllowlist{
```

https://clang.llvm.org/docs/DiagnosticsReference.html#wglobal-constructors while OSS uses `-Wno-global-constructors`.  Let me check with the team on how to deal with this issue. cc @malfet @dagitses

**Comment by fakeYan** [comment]
> From the internal signals, I think there is an discrepancy build issue where `-Werror,-Wglobal-constructors` is set:
> 
> ```
> torch/csrc/jit/serialization/pickler.h:305:44: error: declaration requires a global destructor [-Werror,-Wglobal-constructors]
> [CONTEXT] static std::unordered_set<c10::DeviceType> DeviceTypeAllowlist{
> ```
> 
> https://clang.llvm.org/docs/DiagnosticsReference.html#wglobal-constructors while OSS uses `-Wno-global-constructors`. Let me check with the team on how to deal with this issue. cc @malfet @dagitses

Maybe, it is a solution to release this static variable into the function. Like ``

File: test/cpp_extensions/open_registration_extension.cpp [hunk 6]
```diff
@@ -47,6 +49,68 @@ namespace at::native {
...
+    return c10::BackendMeta::clone(ptr);
+  }
+};
+
+// we need to register two functions for serialization
+void for_serialization(const at::Tensor& t, std::unordered_map<std::string, bool>& m) {
+  if (t.unsafeGetTensorImpl()->get_backend_meta_intrusive_ptr() == nullptr) {
+    return;
+  }
+  CustomBackendMetadata* tmeta = dynamic_cast<CustomBackendMetadata*>(t.unsafeGetTensorImpl()->get_backend_meta());
```
Comment (malfet) [inline test/cpp_extensions/open_registration_extension.cpp:70]:
```suggestion
  auto tmeta = dynamic_cast<CustomBackendMetadata*>(t.unsafeGetTensorImpl()->get_backend_meta());
```

File: torch/csrc/jit/serialization/pickler.h [hunk 7]
```diff
@@ -296,8 +297,59 @@ uint64_t getStorageKey(const at::Tensor& tensor);
 // otherwise return false
 bool checkHasValidSetGetState(const std::shared_ptr<c10::ClassType>& cls);
 
-// Return a map of Tensor Metadata for serialization.
-// For now, it only takes care of `conj` and `neg` bit.
+// Declare BackendMeta serialization and deserialization function pointer types.
+using BackendMetaPtr =
+    void (*)(const at::Tensor&, std::unordered_map<std::string, bool>&);
```
Comment (malfet) [inline torch/csrc/jit/serialization/pickler.h:302]: Nit (yes, it adds a bit of an overhead for C functions but allows one to bind some context if necessary)
```suggestion
    std::function<void(const at::Tensor&, std::unordered_map<std::string, bool>&)>;
```

**Comment by malfet** [comment]
@fakeYan `-Wno-global-constructors` allows only trivially constructible objects, like `std::array` or `list_initializer`. If idea is that `DeviceTypeAllowlist`  is immutable list of few elements, than `list_initializer` would do nicely, wouldn't it?

**Comment by fakeYan** [comment]
Hi @huydhn,

I have changed it, please help me to take a look

**Comment by fakeYan** [comment]
Hi @malfet,

Please take a moment to review my code.

**Review by huydhn — APPROVED** [review]

**Comment by huydhn** [comment]
> Could you please take a moment to review my code and help to merge this pr.

Sorry for the delay.  I have double checked with @malfet to confirm that all signals (external, internal) looks good.  Merging is on the way.

**Pull request closed without merging**